src/render/SDL_palette_sw.c

```diff
@@ -0,0 +1,165 @@
+  arising from the use of this software.
+
+  Permission is granted to anyone to use this software for any purpose,
+  including commercial applications, and to alter it and redistribute it
+  freely, subject to the following restrictions:
+
+  1. The origin of this software must not be misrepresented; you must not
+     claim that you wrote the original software. If you use this software
+     in a product, an acknowledgment in the product documentation would be
+     appreciated but is not required.
+  2. Altered source versions must be plainly marked as such, and must not be
+     misrepresented as being the original software.
+  3. This notice may not be removed or altered from any source distribution.
+*/
+#include "../SDL_internal.h"
+
+/* This is the software implementation of the paletted texture support */
+
+#include "SDL_palette_sw_c.h"
+
+
+SDL_SW_PaletteTexture *
+SDL_SW_CreatePaletteTexture(Uint32 format, int w, int h)
+{
+    SDL_SW_PaletteTexture *swdata;
+
+    switch (format) {
+    /*
+    case SDL_PIXELFORMAT_INDEX1LSB:
+    case SDL_PIXELFORMAT_INDEX1MSB:
+    case SDL_PIXELFORMAT_INDEX4LSB:
+    case SDL_PIXELFORMAT_INDEX4MSB:
+    */
+    case SDL_PIXELFORMAT_INDEX8:
+        break;
+    default:
+        SDL_SetError("Unsupported palette format");
+        return NULL;
+    }
+
+    swdata = (SDL_SW_PaletteTexture *) SDL_calloc(1, sizeof(*swdata));
+    if (!swdata) {
+        SDL_OutOfMemory();
+        return NULL;
+    }
+
+    swdata->surface = SDL_CreateRGBSurfaceWithFormat(0, w, h, SDL_BITSPERPIXEL(format), format);
+    if (!swdata->surface) {
+        SDL_SW_DestroyPaletteTexture(swdata);
+        return NULL;
+    }
+
+    swdata->palette = SDL_AllocPalette(256);
+    if (!swdata->palette) {
+        SDL_SW_DestroyPaletteTexture(swdata);
+        return NULL;
+    }
+
+    SDL_SetSurfacePalette(swdata->surface, swdata->palette);
+
+    /* We're all done.. */
+    return (swdata);
+}
+
+int
+SDL_SW_UpdatePaletteTexture(SDL_SW_PaletteTexture* swdata, const SDL_Rect* rect,
+                            const void* pixels, int pitch,
+                            const SDL_Color* colors, int firstcolor, int ncolors)
+{
+    const Uint8 *src;
+    Uint8 *dst;
+    int row;
+    size_t length;
+
+    if (pixels) {
+        if (SDL_LockSurface(swdata->surface) < 0)
+            return -1;
+
+        src = pixels;
+        dst = (Uint8*)swdata->surface->pixels + rect->y * swdata->surface->pitch + rect->x;
+        length = rect->w;
+        for (row = 0; row < rect->h; ++row) {
+            SDL_memcpy(dst, src, length);
+            src += pitch;
+            dst += swdata->surface->pitch;
+        }
+
+        SDL_UnlockSurface(swdata->surface);
+    }
+    if (colors) {
+        SDL_SetPaletteColors(swdata->palette, colors, firstcolor, ncolors);
+    }
+    return 0;
+}
+
+int
+SDL_SW_LockPaletteTexture(SDL_SW_PaletteTexture * swdata, const SDL_Rect * rect,
+                          void **pixels, int *pitch)
```

Comment:
This place and other places in the code: please align pointers to have the same code style. For example, here we align them in the middle while in the `include/SDL_render.h` we align to left.